### ISSUE TITLE
Narrowing does not work for variables typed with type parameters

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#6445]
```ts
interface Base { b }
interface Derived extends Base { d }

declare function isDerived(x: Base): x is Derived;

function f<T extends Base>(x: T, y: T) {
    if (isDerived(x) && isDerived(y)) {
        return x.d === y.d;
    }

    return false;
}
```

Currently, this errors because both `x` and `y` still have type `T`, so they are missing the `d` property.

**Comment by sandersn** [comment]
`getNarrowedType` doesn't check whether the type it's narrowing is a type parameter with a constraint. It just calls `isTypeAssignableTo(Derived, T)`, which is false.

`getNarrowedType` needs to instead check `isTypeAssignableTo(Derived, Base)`, but I'm not sure whether it's appropriate to narrow to `Derived` at that point, or to `T extends Derived`. Probably the latter.

@weswigham, do you have an idea of the right thing to do here?

**Comment by mhegazy** [comment]
this is handled by https://github.com/Microsoft/TypeScript/pull/8010

**Comment by ahejlsberg** [comment]
Fixed in #8010.
